File: _CodeTopics/LeetCode/000208/000208.py
class TrieNode(object):
    def __init__(self, value):
        self.val = value
        self.isleaf = False
        self.children = {}

class Trie(object):

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.root = TrieNode(None)

    def insert(self, word):
        """
        Inserts a word into the trie.
        :type word: str
        :rtype: None
        """
        curr = self.root
        for letter in word:
            if not curr.children.has_key(letter):
                curr.children[letter] = TrieNode(letter)
            curr = curr.children[letter]
        curr.isleaf = True

    def search(self, word):
        """
        Returns if the word is in the trie.
        :type word: str
        :rtype: bool
        """
        curr = self.root
        for letter in word:
            if not curr.children.has_key(letter):
                return False
            curr = curr.children[letter]
        return False if not curr.isleaf else True

    def startsWith(self, prefix):
        """
        Returns if there is any word in the trie that starts with the given prefix.
        :type prefix: str
        :rtype: bool
        """
        curr = self.root
        for letter in prefix:
            if not curr.children.has_key(letter):
                return False
            curr = curr.children[letter]
        return True

# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)

"""
https://leetcode-cn.com/submissions/detail/95263958/

15 / 15 个通过测试用例
状态：通过
执行用时：368 ms
内存消耗：39.1 MB
"""
"""
执行用时：368 ms, 在所有 Python 提交中击败了53.45%的用户
内存消耗：39.1 MB, 在所有 Python 提交中击败了63.64%的用户
"""

File: _CodeTopics/LeetCode/201-400/000208/tran/000208.cpp
class TrieNode {
// 这里如果是private，下面 if (curr->children.empty() || curr->children.find(letter) == curr->children.end())
// 就会报错说 error: 'children' is a private member of 'TrieNode'
// 所以这里先改成 public，后面再写一个用 private 成员，但是有public的get()方法的。
public:
    char val;
    bool isleaf;
    unordered_map<char, TrieNode*> children;
public:
    TrieNode () {
        this->isleaf = false;
        children = {};
    }
    TrieNode (char value) {
        this->val = value;
        this->isleaf = false;
        children = {};
    }
};

class Trie {
private:
    TrieNode* root;
public:
    /** Initialize your data structure here. */
    Trie() {
        this->root = new TrieNode();
    }
    
    /** Inserts a word into the trie. */
    void insert(string word) {
        auto curr = this->root;
        for (auto letter : word) {
            // 这里python的写法可以不用判断map是否为空，但是C++的不能省。。。
            if (curr->children.empty() || curr->children.find(letter) == curr->children.end()) {
                curr->children[letter] = new TrieNode(letter);
            }
            curr = curr->children[letter];
        }
        curr->isleaf = true;
    }
    
    /** Returns if the word is in the trie. */
    bool search(string word) {
        auto curr = this->root;
        for (auto letter : word) {
            if (curr->children.empty() || curr->children.find(letter) == curr->children.end()) {
                return false;
            }
            curr = curr->children[letter];
        }
        if (!curr->isleaf) {
            return false;
        } else {
            return true;
        }
    }
    
    /** Returns if there is any word in the trie that starts with the given prefix. */
    bool startsWith(string prefix) {
        auto curr = this->root;
        for (auto letter : prefix) {
            if (curr->children.empty() || curr->children.find(letter) == curr->children.end()) {
                return false;
            }
            curr = curr->children[letter];
        }
        return true;
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */

/*
https://leetcode-cn.com/submissions/detail/167614468/

15 / 15 个通过测试用例
状态：通过
执行用时: 100 ms
内存消耗: 42 MB

执行用时：100 ms, 在所有 C++ 提交中击败了21.77%的用户
内存消耗：42 MB, 在所有 C++ 提交中击败了88.50%的用户
*/
